File: sdk/ts/src/receipt/grant-resolver.test.ts
import { describe, expect, it } from "vitest";
import {
	type GrantInfo,
	type GrantResolver,
	GroundedOutcome,
	type GroundedPrincipalViolation,
	verifyGroundedPrincipalTier,
} from "./grant-resolver.js";
import { generateKeyPair, signReceipt } from "./signing.js";
import type { AgentReceipt, RiskLevel } from "./types.js";

// --- test helpers ---

function makeGroundedReceipt(
	principalId: string,
	riskLevel: RiskLevel,
	grantRef: string | undefined,
): AgentReceipt {
	const unsigned = {
		"@context": [
			"https://www.w3.org/ns/credentials/v2",
			"https://agentreceipts.ai/context/v2",
		] as const,
		id: "urn:receipt:00000000-0000-0000-0000-000000000001",
		type: ["VerifiableCredential", "AgentReceipt"] as const,
		version: "0.5.0",
		issuer: { id: "did:key:z6Mk1" },
		issuanceDate: new Date().toISOString(),
		credentialSubject: {
			principal: { id: principalId },
			action: {
				id: "act_00000000-0000-0000-0000-000000000001",
				type: "filesystem.file.write",
				risk_level: riskLevel,
				timestamp: new Date().toISOString(),
			},
			outcome: { status: "success" as const },
			chain: {
				sequence: 1,
				previous_receipt_hash: null,
				chain_id: "chain-grounded-1",
			},
			...(grantRef !== undefined
				? {
						authorization: {
							scopes: ["files:write"],
							granted_at: new Date().toISOString(),
							grant_ref: grantRef,
						},
					}
				: {}),
		},
	};
	const { privateKey } = generateKeyPair();
	return signReceipt(unsigned, privateKey, "did:key:z6Mk1#key-1");
}

class StubResolver implements GrantResolver {
	constructor(
		private readonly grants: Map<string, GrantInfo>,
		private readonly errorRefs: Set<string> = new Set(),
	) {}

	resolveGrant(grantRef: string, _principalId: string): Promise<GrantInfo> {
		if (this.errorRefs.has(grantRef)) {
			return Promise.reject(new Error("resolver: grant not found"));
		}
		const g = this.grants.get(grantRef);
		if (g === undefined) {
			return Promise.reject(new Error("resolver: unknown grant ref"));
		}
		return Promise.resolve(g);
	}
}

// --- tests ---

describe("verifyGroundedPrincipalTier", () => {
	it("returns no violations when resolver is null (base-tier verifier)", async () => {
		const r = makeGroundedReceipt("did:user:alice", "high", undefined);
		const violations = await verifyGroundedPrincipalTier([r], null);
		expect(violations).toHaveLength(0);
	});

	it("skips low and medium receipts even with a resolver configured", async () => {
		const resolver = new StubResolver(new Map());
		const low = makeGroundedReceipt("did:user:alice", "low", undefined);
		const med = makeGroundedReceipt("did:user:alice", "medium", undefined);
		const violations = await verifyGroundedPrincipalTier([low, med], resolver);
		expect(violations).toHaveLength(0);
	});

	it("returns UNGROUNDED_PRINCIPAL for a high receipt with no grant_ref", async () => {
		const resolver = new StubResolver(new Map());
		const r = makeGroundedReceipt("did:user:alice", "high", undefined);
		const violations = await verifyGroundedPrincipalTier([r], resolver);
		expect(violations).toHaveLength(1);
		expect(violations[0]?.outcome).toBe(GroundedOutcome.UngroundedPrincipal);
	});

	it("returns UNGROUNDED_PRINCIPAL for a high receipt with empty grant_ref", async () => {
		const resolver = new StubResolver(new Map());
		const r = makeGroundedReceipt("did:user:alice", "high", "");
		const violations = await verifyGroundedPrincipalTier([r], resolver);
		expect(violations).toHaveLength(1);
		expect(violations[0]?.outcome).toBe(GroundedOutcome.UngroundedPrincipal);
	});

	it("returns UNGROUNDED_PRINCIPAL when the resolver rejects for the grant_ref", async () => {
		const resolver = new StubResolver(new Map(), new Set(["bad-ref"]));
		const r = makeGroundedReceipt("did:user:alice", "high", "bad-ref");
		const violations = await verifyGroundedPrincipalTier([r], resolver);
		expect(violations).toHaveLength(1);
		expect(violations[0]?.outcome).toBe(GroundedOutcome.UngroundedPrincipal);
	});

	it("returns PRINCIPAL_GRANT_MISMATCH when resolved subject does not equal principal.id", async () => {
		const resolver = new StubResolver(
			new Map([
				["grant-abc", { subject: "did:user:bob", scopes: ["files:write"] }],
			]),
		);
		const r = makeGroundedReceipt("did:user:alice", "high", "grant-abc");
		const violations = await verifyGroundedPrincipalTier([r], resolver);
		expect(violations).toHaveLength(1);
		expect(violations[0]?.outcome).toBe(GroundedOutcome.PrincipalGrantMismatch);
	});

	it("returns no violation when the grant resolves and subject equals principal.id", async () => {
		const resolver = new StubResolver(
			new Map([
				["grant-xyz", { subject: "did:user:alice", scopes: ["files:write"] }],
			]),
		);
		const r = makeGroundedReceipt("did:user:alice", "high", "grant-xyz");
		const violations = await verifyGroundedPrincipalTier([r], resolver);
		expect(violations).toHaveLength(0);
	});

	it("checks critical receipts alongside high receipts", async () => {
		const resolver = new StubResolver(new Map());
		const r = makeGroundedReceipt("did:user:alice", "critical", undefined);
		const violations = await verifyGroundedPrincipalTier([r], resolver);
		expect(violations).toHaveLength(1);
		expect(violations[0]?.outcome).toBe(GroundedOutcome.UngroundedPrincipal);
	});

	it("collects all violations rather than stopping at the first", async () => {
		const resolver = new StubResolver(
			new Map([["grant-good", { subject: "did:user:alice", scopes: [] }]]),
			new Set(["grant-bad"]),
		);
		const r1 = makeGroundedReceipt("did:user:alice", "high", undefined); // UNGROUNDED
		const r2 = makeGroundedReceipt("did:user:alice", "high", "grant-good"); // OK
		const r3 = makeGroundedReceipt("did:user:alice", "critical", "grant-bad"); // UNGROUNDED (resolve fails)
		const violations = await verifyGroundedPrincipalTier(
			[r1, r2, r3],
			resolver,
		);
		expect(violations).toHaveLength(2);
	});

	it("reports the correct receipt index in the violation", async () => {
		const resolver = new StubResolver(new Map());
		const low = makeGroundedReceipt("did:user:alice", "low", undefined); // index 0, skipped
		const high = makeGroundedReceipt("did:user:alice", "high", undefined); // index 1, violation
		const violations = await verifyGroundedPrincipalTier([low, high], resolver);
		expect(violations).toHaveLength(1);
		expect(violations[0]?.index).toBe(1);
	});

	it("returns no violations for an empty receipt slice", async () => {
		const resolver = new StubResolver(new Map());
		const violations = await verifyGroundedPrincipalTier([], resolver);
		expect(violations).toHaveLength(0);
	});
});

// Type-level smoke test: ensure the exported types satisfy expected shapes.
const _typeCheck: GroundedPrincipalViolation = {
	index: 0,
	receiptId: "urn:receipt:00000000-0000-0000-0000-000000000001",
	principalId: "did:user:alice",
	grantRef: "grant-abc",
	outcome: GroundedOutcome.UngroundedPrincipal,
	detail: "authorization.grant_ref is absent on a high receipt",
};
void _typeCheck;

File: sdk/go/receipt/grant_resolver_test.go
package receipt

import (
	"errors"
	"testing"
	"time"
)

// stubResolver is a GrantResolver used in tests. It maps grant_ref to a
// pre-built GrantInfo (or an error if the key maps to an error sentinel).
type stubResolver struct {
	grants map[string]GrantInfo
	// errorRefs maps grant_ref strings that should return an error.
	errorRefs map[string]struct{}
}

func (s *stubResolver) ResolveGrant(grantRef, _ string) (GrantInfo, error) {
	if _, bad := s.errorRefs[grantRef]; bad {
		return GrantInfo{}, errors.New("resolver: grant not found")
	}
	if g, ok := s.grants[grantRef]; ok {
		return g, nil
	}
	return GrantInfo{}, errors.New("resolver: unknown grant ref")
}

func makeGroundedReceipt(t *testing.T, kp KeyPair, principalID string, riskLevel RiskLevel, grantRef string) AgentReceipt {
	t.Helper()

	var auth *Authorization
	if grantRef != "" {
		auth = &Authorization{
			Scopes:    []string{"files:write"},
			GrantedAt: time.Now().UTC().Format(time.RFC3339),
			GrantRef:  grantRef,
		}
	}

	unsigned := Create(CreateInput{
		Issuer:        Issuer{ID: "did:key:z6Mk1"},
		Principal:     Principal{ID: principalID},
		Action:        Action{Type: "filesystem.file.write", RiskLevel: riskLevel},
		Outcome:       Outcome{Status: StatusSuccess},
		Authorization: auth,
		Chain:         Chain{Sequence: 1, PreviousReceiptHash: nil, ChainID: "chain-grounded-1"},
	})
	signed, err := Sign(unsigned, kp.PrivateKey, "did:key:z6Mk1#key-1")
	if err != nil {
		t.Fatal(err)
	}
	return signed
}

// TestGrantInfoFields verifies the GrantInfo struct carries the expected fields.
func TestGrantInfoFields(t *testing.T) {
	now := time.Now().UTC()
	g := GrantInfo{
		Subject:   "did:user:alice",
		Scopes:    []string{"files:write"},
		IssuedAt:  now,
		ExpiresAt: now.Add(time.Hour),
		Issuer:    "https://auth.example.com",
	}
	if g.Subject != "did:user:alice" {
		t.Errorf("unexpected Subject: %s", g.Subject)
	}
	if len(g.Scopes) != 1 || g.Scopes[0] != "files:write" {
		t.Errorf("unexpected Scopes: %v", g.Scopes)
	}
}

// TestVerifyGroundedPrincipalTier_NoResolver returns no violations when no
// resolver is configured (base-tier verifier).
func TestVerifyGroundedPrincipalTier_NoResolver(t *testing.T) {
	kp, _ := GenerateKeyPair()
	r := makeGroundedReceipt(t, kp, "did:user:alice", RiskHigh, "")
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{r}, nil)
	if len(violations) != 0 {
		t.Errorf("expected no violations without resolver, got %d", len(violations))
	}
}

// TestVerifyGroundedPrincipalTier_LowMediumSkipped confirms low/medium receipts
// are not checked even when a resolver is configured.
func TestVerifyGroundedPrincipalTier_LowMediumSkipped(t *testing.T) {
	kp, _ := GenerateKeyPair()
	resolver := &stubResolver{grants: map[string]GrantInfo{}}
	lowR := makeGroundedReceipt(t, kp, "did:user:alice", RiskLow, "")
	medR := makeGroundedReceipt(t, kp, "did:user:alice", RiskMedium, "")
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{lowR, medR}, resolver)
	if len(violations) != 0 {
		t.Errorf("expected no violations for low/medium receipts, got %d", len(violations))
	}
}

// TestVerifyGroundedPrincipalTier_MissingGrantRef produces UNGROUNDED_PRINCIPAL
// for a high receipt with no grant_ref.
func TestVerifyGroundedPrincipalTier_MissingGrantRef(t *testing.T) {
	kp, _ := GenerateKeyPair()
	resolver := &stubResolver{grants: map[string]GrantInfo{}}
	r := makeGroundedReceipt(t, kp, "did:user:alice", RiskHigh, "")
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{r}, resolver)
	if len(violations) != 1 {
		t.Fatalf("expected 1 violation, got %d", len(violations))
	}
	if violations[0].Outcome != GroundedOutcomeUngrounded {
		t.Errorf("expected UNGROUNDED_PRINCIPAL, got %s", violations[0].Outcome)
	}
}

// TestVerifyGroundedPrincipalTier_ResolutionFailure produces UNGROUNDED_PRINCIPAL
// when the resolver cannot resolve the grant_ref.
func TestVerifyGroundedPrincipalTier_ResolutionFailure(t *testing.T) {
	kp, _ := GenerateKeyPair()
	resolver := &stubResolver{
		grants:    map[string]GrantInfo{},
		errorRefs: map[string]struct{}{"bad-ref": {}},
	}
	r := makeGroundedReceipt(t, kp, "did:user:alice", RiskHigh, "bad-ref")
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{r}, resolver)
	if len(violations) != 1 {
		t.Fatalf("expected 1 violation, got %d", len(violations))
	}
	if violations[0].Outcome != GroundedOutcomeUngrounded {
		t.Errorf("expected UNGROUNDED_PRINCIPAL, got %s", violations[0].Outcome)
	}
}

// TestVerifyGroundedPrincipalTier_PrincipalMismatch produces PRINCIPAL_GRANT_MISMATCH
// when the resolved grant subject does not equal principal.id.
func TestVerifyGroundedPrincipalTier_PrincipalMismatch(t *testing.T) {
	kp, _ := GenerateKeyPair()
	resolver := &stubResolver{
		grants: map[string]GrantInfo{
			"grant-abc": {
				Subject: "did:user:bob", // mismatch: receipt says alice
				Scopes:  []string{"files:write"},
			},
		},
	}
	r := makeGroundedReceipt(t, kp, "did:user:alice", RiskHigh, "grant-abc")
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{r}, resolver)
	if len(violations) != 1 {
		t.Fatalf("expected 1 violation, got %d", len(violations))
	}
	if violations[0].Outcome != GroundedOutcomeMismatch {
		t.Errorf("expected PRINCIPAL_GRANT_MISMATCH, got %s", violations[0].Outcome)
	}
}

// TestVerifyGroundedPrincipalTier_ValidGrounding produces no violation when
// grant resolves and subject equals principal.id.
func TestVerifyGroundedPrincipalTier_ValidGrounding(t *testing.T) {
	kp, _ := GenerateKeyPair()
	resolver := &stubResolver{
		grants: map[string]GrantInfo{
			"grant-xyz": {
				Subject: "did:user:alice",
				Scopes:  []string{"files:write"},
			},
		},
	}
	r := makeGroundedReceipt(t, kp, "did:user:alice", RiskHigh, "grant-xyz")
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{r}, resolver)
	if len(violations) != 0 {
		t.Errorf("expected no violations for grounded receipt, got %d", len(violations))
	}
}

// TestVerifyGroundedPrincipalTier_CriticalAlsoChecked verifies critical receipts
// are included in the tier checks alongside high receipts.
func TestVerifyGroundedPrincipalTier_CriticalAlsoChecked(t *testing.T) {
	kp, _ := GenerateKeyPair()
	resolver := &stubResolver{grants: map[string]GrantInfo{}}
	r := makeGroundedReceipt(t, kp, "did:user:alice", RiskCritical, "")
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{r}, resolver)
	if len(violations) != 1 {
		t.Fatalf("expected 1 violation for critical receipt without grant_ref, got %d", len(violations))
	}
	if violations[0].Outcome != GroundedOutcomeUngrounded {
		t.Errorf("expected UNGROUNDED_PRINCIPAL, got %s", violations[0].Outcome)
	}
}

// TestVerifyGroundedPrincipalTier_MultipleViolations surfaces all violations,
// not just the first.
func TestVerifyGroundedPrincipalTier_MultipleViolations(t *testing.T) {
	kp, _ := GenerateKeyPair()
	resolver := &stubResolver{
		grants: map[string]GrantInfo{
			"grant-good": {Subject: "did:user:alice"},
		},
		errorRefs: map[string]struct{}{"grant-bad": {}},
	}
	r1 := makeGroundedReceipt(t, kp, "did:user:alice", RiskHigh, "")              // UNGROUNDED
	r2 := makeGroundedReceipt(t, kp, "did:user:alice", RiskHigh, "grant-good")    // OK
	r3 := makeGroundedReceipt(t, kp, "did:user:alice", RiskCritical, "grant-bad") // UNGROUNDED (resolution fails)
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{r1, r2, r3}, resolver)
	if len(violations) != 2 {
		t.Errorf("expected 2 violations, got %d", len(violations))
	}
}

// TestVerifyGroundedPrincipalTier_ReceiptIndexReported confirms the violation
// records the correct receipt index.
func TestVerifyGroundedPrincipalTier_ReceiptIndexReported(t *testing.T) {
	kp, _ := GenerateKeyPair()
	resolver := &stubResolver{grants: map[string]GrantInfo{}}
	low := makeGroundedReceipt(t, kp, "did:user:alice", RiskLow, "")   // index 0, skipped
	high := makeGroundedReceipt(t, kp, "did:user:alice", RiskHigh, "") // index 1, violation
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{low, high}, resolver)
	if len(violations) != 1 {
		t.Fatalf("expected 1 violation, got %d", len(violations))
	}
	if violations[0].Index != 1 {
		t.Errorf("expected violation at index 1, got %d", violations[0].Index)
	}
}

// TestVerifyGroundedPrincipalTier_EmptyChain produces no violations for an
// empty receipt slice.
func TestVerifyGroundedPrincipalTier_EmptyChain(t *testing.T) {
	resolver := &stubResolver{grants: map[string]GrantInfo{}}
	violations := VerifyGroundedPrincipalTier(nil, resolver)
	if len(violations) != 0 {
		t.Errorf("expected no violations for empty chain, got %d", len(violations))
	}
}

// TestVerifyGroundedPrincipalTier_TypedNilResolver confirms that a typed-nil
// resolver (a non-nil interface value whose concrete pointer is nil) is treated
// the same as a nil interface — no violations, no panic.
func TestVerifyGroundedPrincipalTier_TypedNilResolver(t *testing.T) {
	kp, _ := GenerateKeyPair()
	r := makeGroundedReceipt(t, kp, "did:user:alice", RiskHigh, "")

	var typed *stubResolver // typed nil: concrete type known, pointer is nil
	violations := VerifyGroundedPrincipalTier([]AgentReceipt{r}, typed)
	if len(violations) != 0 {
		t.Errorf("expected no violations for typed-nil resolver, got %d", len(violations))
	}
}
